lib/cache/collection.go

```diff
@@ -0,0 +1,195 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cache
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+// collection is responsible for managing a cached resource.
+type collection[T any] struct {
+	getAll          func(ctx context.Context, loadSecrets bool) ([]T, error)
+	store           *resourceStore[T]
+	watch           types.WatchKind
+	headerTransform func(hdr *types.ResourceHeader) T
+	filter          func(T) bool
+	singleton       bool
+}
```

Comment:
nit: the behavior of these fields should have fairly detailed documentation since this is definitely something that'll be tricky to parse for someone new to caching concepts.  Also, it might be more clear to name `getAll` something like `fetch` or `fetcher` since we already have a well-established nomenclature for that behavior.